src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java

```diff
@@ -93,42 +111,59 @@ private void createColumns(UUID tableId, Collection<Column> columns) {
   }
 
   public List<SnapshotTable> retrieveTables(UUID parentId) {
-    MapSqlParameterSource params = new MapSqlParameterSource().addValue("parent_id", parentId);
-    return jdbcTemplate.query(
+    MapSqlParameterSource params = new MapSqlParameterSource().addValue("snapshot_id", parentId);
+    Map<UUID, SnapshotTable> tableMap = new TreeMap<>();
+    jdbcTemplate.query(
         sqlSelectTable,
         params,
         (rs, rowNum) -> {
+          UUID tableId = rs.getObject("table_id", UUID.class);
           SnapshotTable table =
-              new SnapshotTable()
-                  .id(rs.getObject("id", UUID.class))
-                  .name(rs.getString("name"))
-                  .rowCount(rs.getLong("row_count"));
-          List<Column> columns = retrieveColumns(table);
-          table.columns(columns);
+              tableMap.computeIfAbsent(
+                  tableId,
+                  (id) -> {
+                    try {
+                      return new SnapshotTable()
+                          .id(id)
+                          .name(rs.getString("table_name"))
+                          .rowCount(rs.getLong("table_row_count"))
+                          .columns(new ArrayList<>())
+                          .primaryKey(new ArrayList<>());
+                    } catch (SQLException e) {
+                      throw new RuntimeException(e);
+                    }
+                  });
 
-          Map<String, Column> columnMap =
-              columns.stream().collect(Collectors.toMap(Column::getName, Function.identity()));
+          // Add column to table
+          Column column = getColumnRowMapper(table).mapRow(rs, rowNum);
+          table.getColumns().add(column);
 
-          List<String> primaryKey = DaoUtils.getStringList(rs, "primary_key");
-          List<Column> naturalKeyColumns =
-              primaryKey.stream().map(columnMap::get).collect(Collectors.toList());
-          table.primaryKey(naturalKeyColumns);
+          // Add primary key to table if appropriate
+          List<String> primaryKey = DaoUtils.getStringList(rs, "table_primary_key");
+          if (primaryKey.contains(column.getName())) {
+            table.getPrimaryKey().add(column);
+          }
 
           return table;
         });
+    return List.copyOf(tableMap.values());
```

Comment:
It's a minor thing but we use `List` in many places where `Collection` would be fine, and this is probably one of those cases. Looking at all callers of this method, there are only five cases. Four all use `stream()` and the fifth passes it to `Collections.unmodifiableList()`. If that was changed, you could return `values()` directly.

Reply:
The problem is that collection doesn't preserve order so we explicitly want to indicate that order _is_ deterministic in this case

Reply:
> The problem is that collection doesn't preserve order

I would separate type from behavior here.

`Collection` is not a concrete type so the implementation type may or may not preserve order. If the contract is that the returned object preserves order, then it doesn't matter that the type is `Collection` vs `List`. Note that even collections like `Map` or `Set` can be order preserving even though their APIs don't guarantee it.